### ISSUE TITLE
Enhancement: Extract proper noun vocabulary from corpus

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#227]
## Motivation

Proper nouns (people, places, organizations) are critical for factual Q&A but are not in the core root vocabulary (11,121 content words). Currently, all unknown proper nouns map to a single `<UNK>` token, meaning "Zamenhof", "Einstein", and "Tolkien" get identical embeddings.

## Proposal

Extract a **proper noun vocabulary** from the corpus and give common proper nouns their own learned embeddings.

### Architecture (Aligned with Klareco Philosophy)

**Deterministic extraction + minimal learned parameters:**

```python
# Stage 1: Extract proper nouns (DETERMINISTIC)
proper_nouns = extract_all_proper_nouns(corpus)
# Sources:
# - Wikipedia article titles (2,747 articles)
# - Capitalized words in parsed corpus
# - Filter: require 10+ occurrences to avoid noise
# Result: ~5K-10K proper nouns

# Stage 2: Create vocabulary
proper_noun_vocab = {
    "Zamenhof": 11122,
    "Bialystok": 11123, 
    "Esperanto": 11124,
    # ... 5K-10K entries
}

# Stage 3: Embedding layer
class CompositionalEmbedding:
    def __init__(self, ...):
        # Existing embeddings
        self.root_embed = nn.Embedding(11121, 64)  # Content words
        
        # NEW: Proper noun embeddings
        self.proper_noun_embed = nn.Embedding(10000, 64)  # Top 10K proper nouns
        
    def get_root_idx(self, root: str, is_proper_noun: bool = False):
        if root in self.root_vocab:
            return ('content', self.root_vocab[root])
        
        if is_proper_noun and root in self.proper_noun_vocab:
            return ('proper', self.proper_noun_vocab[root])
        
        if is_proper_noun:
            return ('special', self.special_tokens['<UNK_PROPER>'])
        else:
            return ('special', self.special_tokens['<UNK>'])
```

### Benefits

1. **Common proper nouns get individual embeddings**
   - "Zamenhof" and "Einstein" can have different semantic representations
   - Enables similarity between related people/places

2. **Rare proper nouns fall back gracefully**
   - Unknown person → `<UNK_PROPER>` (distinct from `<UNK>` for unknown content words)
   - Still better than complete failure

3. **Aligns with Klareco philosophy**
   - Deterministic extraction (no randomness)
   - Learn only what's necessary (~640KB for 10K × 64d embeddings)
   - Maximal use of corpus structure

4. **Improves factual Q&A**
   - Parser can extract proper noun subjects (fixes #226)
   - Slot-based matching can use entity information
   - Estimated impact: +10-15pp on factual questions

### Implementation Plan

**Phase 1: Extraction (scripts/extract_proper_nouns.py)**
```python
def extract_proper_nouns(corpus_path: Path) -> Dict[str, int]:
    """Extract proper nouns with occurrence counts."""
    proper_nouns = defaultdict(int)
    
    with open(corpus_path) as f:
        for line in f:
            doc = json.loads(line)
            
            # Method 1: Wikipedia titles
            if doc.get('source') == 'wikipedia':
                title = doc['source_metadata']['article_title']
                proper_nouns[normalize(title)] += doc.get('sentence_count', 1)
            
            # Method 2: Capitalized words in sentences
            ast = doc.get('ast', {})
            for word in extract_words_from_ast(ast):
                if word.get('parse_status') == 'proper_name_unknown':
                    root = word.get('radiko', '').lower()
                    if root and len(root) > 2:  # Filter noise
                        proper_nouns[root] += 1
    
    # Filter: require 10+ occurrences
    return {k: v for k, v in proper_nouns.items() if v >= 10}
```

**Phase 2: Vocabulary Creation**
```bash
python scripts/extract_proper_nouns.py \
  --corpus data/corpus/unified_corpus.jsonl \
  --output data/vocabularies/proper_noun_vocab.json \
  --min-occurrences 10
```

**Phase 3: Integration**
- Modify `CompositionalEmbedding` to load proper noun vocabulary
- Update `get_root_idx()` to use proper noun embeddings
- Modify parser to flag proper nouns (already has `proper_name_unknown` status)

**Phase 4: Retraining**
- Add proper noun embeddings to existing model checkpoint
- Initialize new embeddings randomly (or from content word embeddings)
- Fine-tune on corpus with proper noun coverage

### Parameter Impact

```
Current model:
- Root embeddings: 11,121 × 64 = 711,744 params

With proper nouns:
- Root embeddings: 11,121 × 64 = 711,744 params
- Proper noun embeddings: 10,000 × 64 = 640,000 params
- Total: 1,351,744 params (+90% increase)

But relative to full model (15M params): +4% only
```

### Alternative: Hash-Based Embeddings

For comparison, a hash-based approach would be:
```python
# Hash unknown proper nouns to 1000 buckets
hash_bucket = xxhash.xxh64(root.encode()).intdigest() % 1000
return ('hash', self.vocab_size + hash_bucket)
```

**Rejected because:**
- Adds randomness (conflicts with deterministic philosophy)
- Hash collisions unpredictable
- Can't leverage corpus statistics (frequency, co-occurrence)

### Expected Impact

- Factual Q&A: 8% → 40%+ accuracy (+32pp)
- Combined with #225 and #226: 8% → 50%+ accuracy
- Overall benchmark: 38% → 55%+ accuracy

### Related Issues

- #226 - Parser bug for proper nouns (prerequisite)
- #225 - Fix query matching (benefits from this)
- #223, #224 - Diagnostic tools

## Acceptance Criteria

- [ ] Script extracts 5K-10K proper nouns from corpus
- [ ] Vocabulary saved to `data/vocabularies/proper_noun_vocab.json`
- [ ] `CompositionalEmbedding` loads and uses proper noun vocab
- [ ] Parser flags proper nouns (uses existing `proper_name_unknown` status)
- [ ] "Zamenhof" and "Einstein" get different embedding vectors
- [ ] Unknown proper nouns fall back to `<UNK_PROPER>`
- [ ] Factual Q&A accuracy improves measurably

## Priority

**P1 - High**. This is a longer-term enhancement that significantly improves proper noun handling. Should be done after fixing #226 (parser bug).

**Comment by marctjones** [comment]
Proper noun extraction implemented. Part of current parser.
